Support commands against Windows targets in external plugins

MikaelSmith opened the pull request:
Adds support for executing powershell commands instead of POSIX when the `os.login_shell` attribute is set to `powershell`.

Resolves #699.

MikaelSmith committed:
Add loginshell attribute

Adds the `loginshell` attribute, which can be used to determine what
types of shell commands to run on an entry.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Implement ps commands for powershell

Support listing processes on Windows using powershell.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Implement PowerShell support in volume.FS

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Switch to more general OS attribute

Move login shell to a more general OS attribute that may later capture
other information about the operating system on an entry that has one.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented on volume/fs.go:
```diff
@@ -95,7 +95,7 @@ func exec(ctx context.Context, executor plugin.Execable, cmdline []string, tty b
 // VolumeList satisfies the Interface required by List to enumerate files.
 func (d *FS) VolumeList(ctx context.Context, path string) (DirMap, error) {
-	cmdline := StatCmd(path, d.maxdepth)
+	cmdline := d.selectShellCommand(StatCmd(path, d.maxdepth), ChildItemsCmd(path, d.maxdepth))
```
TODO: These changes should have tests associated with them.

ekinanp commented on cmd/ps.go:
```diff
@@ -112,45 +114,80 @@ func parseEntry(line string) (psresult, error) {
 }
 
-func parseLines(node string, chunk string) []psresult {
+func parseStatLines(chunk string) ([]psresult, error) {
```
Might help to re-arrange some of these functions as something like

```
// PS POSIX
<posix stuff>

// PS POWERSHELL
<powershell stuff>
```

so a maintainer can immediately see what's what.

ekinanp commented on cmd/ps.go:
```diff
@@ -183,27 +220,61 @@ func psMain(cmd *cobra.Command, args []string) exitCode {
 }
+
+var dispatchers = []struct {
+	exec  func(client.Client, string) (<-chan apitypes.ExecPacket, error)
```
Nit: `execPS`, `parseOutput`? Slightly clearer.

MikaelSmith commented on plugin/entryAttributes.go:
```diff
@@ -55,6 +85,8 @@ type EntryAttributes struct {
 	ctime   time.Time
 	crtime  time.Time
+	os      OS
+	hasOS   bool
```
Some things will not have any OS information because they're not a computer. I think it makes sense to reflect that directly, rather than having an empty struct.

ekinanp replied:
Also makes sense.

MikaelSmith commented on volume/items.go:
```diff
@@ -0,0 +1,119 @@
+)
+
+// ChildItemsCmd returns the PowerShell command required to stat all the files in a directory up to maxdepth.
+func ChildItemsCmd(path string, maxdepth int) []string {
```
Yeah, I thought about something like that but hadn't decided on names. These work.

MikaelSmith committed:
Make volume stat functions symmetric

Rename stat and items to StatPOSIX and StatPowershell to make it clear
they're related.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Add testing for volume.FS powershell support

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
Updated, added tests for volume.FS.

ekinanp commented on volume/stat.go:
```diff
@@ -1,94 +1,51 @@
-	// List uses "" to mean root. Translate for executing on the target.
-	if path == RootPath {
-		path = "/"
+func addAttributesForPath(dirmap DirMap, attr plugin.EntryAttributes, base, fullpath string, maxdepth int) {
```
Thanks, this helped a bit. I think this could be `addEntryToDirMap`; `attr` => `entryAttr`, `fullpath` => `fullEntryPath` (or `absEntryPath` for consistency?), then `relative` => `relativeEntryPath`

MikaelSmith replied:
But none of the things it deals with are an Entry. Just the path representing one.

ekinanp replied:
Yeah, it's subjective. If I read that, I'd think of it as "addEntryToDirMap" => "where's the entry ...?" => "Oh the path is the entry. Ok this makes sense now".

ekinanp reviewed (approved):
Comment's not a blocker, but I think addressing it would help w/ clarity a bit.

MikaelSmith commented:
Going to add some external plugin docs in a separate PR.
